flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogBase.java

```diff
@@ -197,4 +204,140 @@ protected void alterHiveDatabase(String name, Database newHiveDatabase, boolean
 			throw new CatalogException(String.format("Failed to alter database %s", name), e);
 		}
 	}
+
+	// ------ tables ------
+
+	protected void createHiveTable(ObjectPath tablePath, Table table, boolean ignoreIfExists)
+			throws TableAlreadyExistException, DatabaseNotExistException, CatalogException {
+		if (!databaseExists(tablePath.getDatabaseName())) {
+			throw new DatabaseNotExistException(catalogName, tablePath.getDatabaseName());
+		} else {
+			try {
+				client.createTable(table);
+			} catch (AlreadyExistsException e) {
+				if (!ignoreIfExists) {
+					throw new TableAlreadyExistException(catalogName, tablePath);
+				}
+			} catch (TException e) {
+				throw new CatalogException(String.format("Failed to create table %s", tablePath.getFullName()), e);
+			}
+		}
+	}
+
+	@Override
+	public void renameTable(ObjectPath tablePath, String newTableName, boolean ignoreIfNotExists)
+			throws TableNotExistException, TableAlreadyExistException, CatalogException {
+		try {
+			// alter_table() doesn't throw a clear exception when target table doesn't exist. Thus, check the table existence explicitly
+			if (tableExists(tablePath)) {
+				ObjectPath newPath = new ObjectPath(tablePath.getDatabaseName(), newTableName);
+				// alter_table() doesn't throw a clear exception when new table already exists. Thus, check the table existence explicitly
+				if (tableExists(newPath)) {
+					throw new TableAlreadyExistException(catalogName, newPath);
+				} else {
+					Table table = getHiveTable(tablePath);
+					table.setTableName(newTableName);
+					client.alter_table(tablePath.getDatabaseName(), tablePath.getObjectName(), table);
+				}
+			} else if (!ignoreIfNotExists) {
+				throw new TableNotExistException(catalogName, tablePath);
+			}
+		} catch (TException e) {
+			throw new CatalogException(
+				String.format("Failed to rename table %s", tablePath.getFullName()), e);
+		}
+	}
+
+	protected Table getHiveTable(ObjectPath tablePath) throws TableNotExistException {
+		try {
+			return client.getTable(tablePath.getDatabaseName(), tablePath.getObjectName());
+		} catch (NoSuchObjectException e) {
+			throw new TableNotExistException(catalogName, tablePath);
+		} catch (TException e) {
+			throw new CatalogException(
+				String.format("Failed to get table %s from Hive metastore", tablePath.getFullName()), e);
+		}
+	}
+
+	@Override
+	public void alterTable(ObjectPath tablePath, CatalogBaseTable newTable, boolean ignoreIfNotExists)
+			throws TableNotExistException, CatalogException {
+		if (!tableExists(tablePath)) {
+			if (!ignoreIfNotExists) {
+				throw new TableNotExistException(catalogName, tablePath);
+			}
+		} else {
+			// IMetastoreClient.alter_table() requires the table to have a valid location, which it doesn't in this case
+			// Thus we have to translate alterTable() into (dropTable() + createTable())
+			dropTable(tablePath, false);
```

Comment:
I think we needs to be very careful here. dropTable() might also drop the data, but alterTable() really is supposed to change the metadata only.

We should probably solve the location problem rather than taking this workaround.

Reply:
We also need to note that view and table are sharing this call, so it probably doesn't make sense to change a view to a regular table and vise versa.

Reply:
good catch, fixed